src/measure_theory/lp_space.lean

```diff
@@ -632,4 +640,168 @@ end
+    exact subtype.eq (ae_eq_fun.ext (hf.trans ae_eq_fun.coe_fn_zero.symm)), },
+  { exact absurd hf (snorm_ne_top f), },
+end
+
+lemma coe_fn_neg {f : Lp E p μ} : ⇑(-f) =ᵐ[μ] -f := ae_eq_fun.coe_fn_neg _
+
+@[simp] lemma Lp_norm_neg {f : Lp E p μ} : Lp_norm (-f) = Lp_norm f :=
+by rw [Lp_norm, Lp_norm, snorm_congr_ae coe_fn_neg, snorm_neg]
+
+lemma coe_fn_add {f g : Lp E p μ} : ⇑(f + g) =ᵐ[μ] f + g := ae_eq_fun.coe_fn_add _ _
+
+lemma coe_fn_sub {f g : Lp E p μ} : ⇑(f - g) =ᵐ[μ] f - g := ae_eq_fun.coe_fn_sub _ _
+
+instance : has_dist (Lp E p μ) := { dist := λ f g, Lp_norm (f - g) }
+
+@[simp] lemma dist_eq_Lp_norm_sub {f g : Lp E p μ} : dist f g = Lp_norm (f - g) := rfl
+
+private lemma dist_triangle' [hp : fact (1 ≤ p)] (f g h : Lp E p μ) :
+  dist f h ≤ dist f g + dist g h :=
+begin
+  simp only [dist, Lp_norm],
+  rw ←ennreal.to_real_add (snorm_ne_top (f - g)) (snorm_ne_top (g - h)),
+  suffices h_snorm : snorm ⇑(f - h) p μ ≤ snorm ⇑(f - g) p μ + snorm ⇑(g - h) p μ,
+  { rwa ennreal.to_real_le_to_real (snorm_ne_top (f - h)),
+    exact ennreal.add_ne_top.mpr ⟨snorm_ne_top (f - g), snorm_ne_top (g - h)⟩, },
+  have h_add : (f - h) = f - g + (g - h), by abel,
+  rw [h_add, snorm_congr_ae coe_fn_add],
+  exact snorm_add_le (ae_measurable (f - g)) (ae_measurable (g - h)) hp,
+end
+
+instance [hp : fact (1 ≤ p)] : metric_space (Lp E p μ) :=
+{ dist_self := λ _, by simp,
+  dist_comm := λ _ _, by { simp only [dist], rw [←Lp_norm_neg, neg_sub] },
+  dist_triangle := dist_triangle',
+  eq_of_dist_eq_zero := λ _ _ h, by simpa [Lp_norm_eq_zero_iff (ennreal.zero_lt_one.trans_le hp),
+    sub_eq_zero] using h, }
+
+instance [fact (1 ≤ p)] : normed_group (Lp E p μ) := { dist_eq := λ _ _, by simp [dist, norm] }
+
+section normed_space
+
+variables {𝕜 : Type*} [normed_field 𝕜] [normed_space 𝕜 E]
+
+lemma mem_Lp_const_smul (c : 𝕜) (f : Lp E p μ) : c • f.val ∈ Lp E p μ :=
+begin
+  rw [mem_Lp_iff_snorm_lt_top, snorm_congr_ae (ae_eq_fun.coe_fn_smul _ _), snorm_const_smul,
+    ennreal.mul_lt_top_iff],
+  exact or.inl ⟨ennreal.coe_lt_top, f.prop⟩,
+end
+
+instance : has_scalar 𝕜 (Lp E p μ) := { smul := λ c f, ⟨c • f.val, mem_Lp_const_smul c f⟩ }
+
+lemma coe_fn_smul {f : Lp E p μ} {c : 𝕜} : ⇑(c • f) =ᵐ[μ] c • f := ae_eq_fun.coe_fn_smul _ _
+
+@[simp] lemma Lp_norm_const_smul (c : 𝕜) (f : Lp E p μ) : Lp_norm (c • f) = ∥c∥ * Lp_norm f :=
+by rw [Lp_norm, snorm_congr_ae coe_fn_smul, snorm_const_smul c, ennreal.to_real_mul,  Lp_norm,
+    ennreal.coe_to_real, coe_nnnorm]
+
+instance : mul_action 𝕜 (Lp E p μ) :=
+{ one_smul := λ _, subtype.eq (one_smul 𝕜 _),
+  mul_smul := λ _ _ _, subtype.eq (mul_smul _ _ _) }
+
+instance : distrib_mul_action 𝕜 (Lp E p μ) :=
+{ smul_add := λ _ _ _, subtype.eq (smul_add _ _ _),
+  smul_zero := λ _, subtype.eq (smul_zero _) }
+
+instance : semimodule 𝕜 (Lp E p μ) :=
+{ add_smul := λ _ _ _, subtype.eq (add_smul _ _ _),
+  zero_smul := λ _, subtype.eq (zero_smul _ _) }
+
+instance : vector_space 𝕜 (Lp E p μ) := infer_instance
```

Comment:
This one is not necessary, as it is exactly the same as the semimodule instance. Also, it is not necessary to state all the instances (as each of them is implied by the next one): this will create additional paths for typeclass inference, and may slow down some searches. I'd go just for semimodule and normed_space (removing mul_action and distrib_mul_action).